moveit_ros/moveit_servo/include/moveit_servo/servo.hpp

```diff
@@ -0,0 +1,171 @@
+#include <moveit/planning_scene_monitor/planning_scene_monitor.h>
+#include <moveit_servo/collision_monitor.hpp>
+#include <moveit_servo/command_processor.hpp>
+#include <moveit_servo/datatypes.hpp>
+#include <moveit_servo/status_codes.hpp>
+#include <moveit_servo_lib_parameters.hpp>
+
+namespace moveit_servo
+{
+
+class Servo
+{
+public:
+  Servo(const rclcpp::Node::SharedPtr& node, std::shared_ptr<const servo::ParamListener>& servo_param_listener);
+
+  /**
+   * \brief Computes the required joint state to follow the given command.
+   * @param command The command to follow, std::variant type, can handle JointJog, Twist and Pose.
+   * @return The required joint state.
+   */
+  KinematicState getNextJointState(const ServoInput& command);
+
+  /**
+   * \brief Compute the change in joint position for the received command.
+   * @param command The incoming servo command.
+   * @return The joint position change required (delta).
+   */
+  Eigen::VectorXd jointDeltaFromCommand(const ServoInput& command);
+
+  /**
+   * \brief Set the type of incoming servo command.
+   * @param command_type The type of command servo should expect.
+   */
+  void incomingCommandType(const CommandType& command_type);
+
+  /**
+   * \brief Get the type of command that servo is currently expecting.
+   * @return The type of command.
+   */
+  CommandType incomingCommandType();
+
+  /**
+   * \brief Get the current status of servo.
+   * @return The current status.
+   */
+  StatusCode getStatus();
+
+  /**
+   * \brief Get the message associated with the current servo status.
+   * @return The status message.
+   */
+  const std::string getStatusMessage();
+
+private:
+  /**
+   * \brief Validate the servo parameters
+   * @param servo_params The servo parameters
+   */
+  void validateParams(const servo::Params& servo_params);
+
+  /**
+   * \brief Creates the planning scene monitor used by servo
+   */
+  void createPlanningSceneMonitor();
+
+  /**
+   * \brief create and initialize the smoothing plugin to be used by servo.
+   */
+  void setSmoothingPlugin();
+
+  /**
+   * \brief The callback for velocity scaling values from collision checker.
+   */
+  void collisionVelocityScaleCB(const std_msgs::msg::Float64::ConstSharedPtr& msg);
+
+  /**
+   * \brief Updates the servo parameters and performs some validations.
+   */
+  void updateParams();
+
+  /**
+   * \brief Apply halting logic to specified joints.
+   * @param joints_to_halt The indices of joints to be halted.
+   * @param current_state The current kinematic state.
+   * @param target_state The target kinematic state.
+   * @return The bounded kinematic state.
+   */
+  KinematicState haltJoints(const std::vector<int>& joints_to_halt, const KinematicState& current_state,
+                            const KinematicState& target_state);
+
+  // Variables
+
+  const rclcpp::Node::SharedPtr node_;
+  std::unique_ptr<CommandProcessor> command_processor_;
+  std::atomic<CommandType> incoming_command_type_;
```

Comment:
Is this accurate?

```suggestion
  // This needs to be threadsafe so it can be updated in realtime with Dynamic Reconfigure
  std::atomic<CommandType> incoming_command_type_;
```